// @capacitor/electron/dist/esm/electron/splashscreen.js
import { __assign, __awaiter, __extends, __generator } from "tslib";
import { WebPlugin } from "@capacitor/core";
var SplashScreenPluginElectron = /** @class */ (function (_super) {
    __extends(SplashScreenPluginElectron, _super);
    function SplashScreenPluginElectron() {
        var _this = _super.call(this, {
            name: 'SplashScreen',
            platforms: ['electron']
        }) || this;
        _this.ipc = null;
        _this.ipc = require('electron').ipcRenderer;
        return _this;
    }
    SplashScreenPluginElectron.prototype.show = function (options, callback) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.ipc.send('showCapacitorSplashScreen', __assign({}, options));
                if (callback) {
                    callback();
                }
                return [2 /*return*/];
            });
        });
    };
    SplashScreenPluginElectron.prototype.hide = function (options, callback) {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                this.ipc.send('hideCapacitorSplashScreen', __assign({}, options));
                if (callback) {
                    callback();
                }
                return [2 /*return*/];
            });
        });
    };
    return SplashScreenPluginElectron;
}(WebPlugin));
export { SplashScreenPluginElectron };
var SplashScreen = new SplashScreenPluginElectron();
export { SplashScreen };
