// ngx-stripe/esm5/lib/interfaces/element.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function Element() { }
if (false) {
    /**
     * @param {?} el
     * @return {?}
     */
    Element.prototype.mount = function (el) { };
    /**
     * @param {?} ev
     * @param {?} handler
     * @return {?}
     */
    Element.prototype.on = function (ev, handler) { };
    /**
     * @return {?}
     */
    Element.prototype.blur = function () { };
    /**
     * @return {?}
     */
    Element.prototype.clear = function () { };
    /**
     * @return {?}
     */
    Element.prototype.focus = function () { };
    /**
     * @return {?}
     */
    Element.prototype.unmount = function () { };
    /**
     * @param {?} options
     * @return {?}
     */
    Element.prototype.update = function (options) { };
}
/**
 * @record
 */
export function ElementOptions() { }
if (false) {
    /** @type {?|undefined} */
    ElementOptions.prototype.style;
    /** @type {?|undefined} */
    ElementOptions.prototype.hidePostalCode;
    /** @type {?|undefined} */
    ElementOptions.prototype.supportedCountries;
    /** @type {?|undefined} */
    ElementOptions.prototype.hideIcon;
    /** @type {?|undefined} */
    ElementOptions.prototype.iconStyle;
    /** @type {?|undefined} */
    ElementOptions.prototype.placeholder;
    /** @type {?|undefined} */
    ElementOptions.prototype.value;
    /** @type {?|undefined} */
    ElementOptions.prototype.placeholderCountry;
}
/**
 * @record
 */
export function ElementStyleAttributes() { }
if (false) {
    /** @type {?|undefined} */
    ElementStyleAttributes.prototype.color;
    /** @type {?|undefined} */
    ElementStyleAttributes.prototype.fontFamily;
    /** @type {?|undefined} */
    ElementStyleAttributes.prototype.fontSize;
    /** @type {?|undefined} */
    ElementStyleAttributes.prototype.fontSmoothing;
    /** @type {?|undefined} */
    ElementStyleAttributes.prototype.fontStyle;
    /** @type {?|undefined} */
    ElementStyleAttributes.prototype.fontWeight;
    /** @type {?|undefined} */
    ElementStyleAttributes.prototype.fontVariant;
    /** @type {?|undefined} */
    ElementStyleAttributes.prototype.iconColor;
    /** @type {?|undefined} */
    ElementStyleAttributes.prototype.lineHeight;
    /** @type {?|undefined} */
    ElementStyleAttributes.prototype.letterSpacing;
    /** @type {?|undefined} */
    ElementStyleAttributes.prototype.textDecoration;
    /** @type {?|undefined} */
    ElementStyleAttributes.prototype.textShadow;
    /** @type {?|undefined} */
    ElementStyleAttributes.prototype.textTransform;
    /* Skipping unnamed member:
    ':hover'?: any;*/
    /* Skipping unnamed member:
    ':focus'?: any;*/
    /* Skipping unnamed member:
    '::placeholder'?: any;*/
    /* Skipping unnamed member:
    '::selection'?: any;*/
    /* Skipping unnamed member:
    ':-webkit-autofill'?: any;*/
}
