// ngx-plyr/esm2015/lib/plyr-driver/plyr-driver.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function PlyrDriverCreateParams() { }
if (false) {
    /** @type {?} */
    PlyrDriverCreateParams.prototype.options;
    /** @type {?} */
    PlyrDriverCreateParams.prototype.videoElement;
}
/**
 * @record
 */
export function PlyrDriverUpdateSourceParams() { }
if (false) {
    /** @type {?} */
    PlyrDriverUpdateSourceParams.prototype.plyr;
    /** @type {?} */
    PlyrDriverUpdateSourceParams.prototype.source;
    /** @type {?} */
    PlyrDriverUpdateSourceParams.prototype.videoElement;
}
/**
 * @record
 */
export function PlyrDriverDestroyParams() { }
if (false) {
    /** @type {?} */
    PlyrDriverDestroyParams.prototype.plyr;
}
/**
 * @record
 */
export function PlyrDriver() { }
if (false) {
    /**
     * @param {?} params
     * @return {?}
     */
    PlyrDriver.prototype.create = function (params) { };
    /**
     * @param {?} params
     * @return {?}
     */
    PlyrDriver.prototype.updateSource = function (params) { };
    /**
     * @param {?} params
     * @return {?}
     */
    PlyrDriver.prototype.destroy = function (params) { };
}
