// ngx-stripe/esm2015/lib/interfaces/token.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @record
 */
export function Token() { }
if (false) {
    /** @type {?} */
    Token.prototype.id;
    /** @type {?} */
    Token.prototype.object;
    /** @type {?|undefined} */
    Token.prototype.bank_account;
    /** @type {?|undefined} */
    Token.prototype.card;
    /** @type {?} */
    Token.prototype.client_ip;
    /** @type {?} */
    Token.prototype.livemode;
    /** @type {?} */
    Token.prototype.type;
    /** @type {?} */
    Token.prototype.used;
}
/**
 * @record
 */
export function CardDataOptions() { }
if (false) {
    /** @type {?|undefined} */
    CardDataOptions.prototype.name;
    /** @type {?|undefined} */
    CardDataOptions.prototype.address_line1;
    /** @type {?|undefined} */
    CardDataOptions.prototype.address_line2;
    /** @type {?|undefined} */
    CardDataOptions.prototype.address_city;
    /** @type {?|undefined} */
    CardDataOptions.prototype.address_state;
    /** @type {?|undefined} */
    CardDataOptions.prototype.address_zip;
    /** @type {?|undefined} */
    CardDataOptions.prototype.address_country;
    /** @type {?|undefined} */
    CardDataOptions.prototype.currency;
}
/**
 * @record
 */
export function TokenResult() { }
if (false) {
    /** @type {?|undefined} */
    TokenResult.prototype.token;
    /** @type {?|undefined} */
    TokenResult.prototype.error;
}
/**
 * @record
 */
export function Address() { }
if (false) {
    /** @type {?} */
    Address.prototype.line1;
    /** @type {?|undefined} */
    Address.prototype.line2;
    /** @type {?} */
    Address.prototype.city;
    /** @type {?} */
    Address.prototype.state;
    /** @type {?} */
    Address.prototype.postal_code;
}
/**
 * @record
 */
export function DateOfBirth() { }
if (false) {
    /** @type {?} */
    DateOfBirth.prototype.day;
    /** @type {?} */
    DateOfBirth.prototype.month;
    /** @type {?} */
    DateOfBirth.prototype.year;
}
/**
 * @record
 */
export function LegalEntity() { }
if (false) {
    /** @type {?} */
    LegalEntity.prototype.address;
    /** @type {?|undefined} */
    LegalEntity.prototype.address_kana;
    /** @type {?|undefined} */
    LegalEntity.prototype.address_kanji;
    /** @type {?|undefined} */
    LegalEntity.prototype.dob;
    /** @type {?} */
    LegalEntity.prototype.first_name;
    /** @type {?|undefined} */
    LegalEntity.prototype.first_name_kana;
    /** @type {?|undefined} */
    LegalEntity.prototype.first_name_kanji;
    /** @type {?|undefined} */
    LegalEntity.prototype.gender;
    /** @type {?} */
    LegalEntity.prototype.last_name;
    /** @type {?|undefined} */
    LegalEntity.prototype.last_name_kana;
    /** @type {?|undefined} */
    LegalEntity.prototype.last_name_kanji;
    /** @type {?|undefined} */
    LegalEntity.prototype.maiden_name;
    /** @type {?|undefined} */
    LegalEntity.prototype.personal_id_number;
    /** @type {?|undefined} */
    LegalEntity.prototype.phone_number;
    /** @type {?|undefined} */
    LegalEntity.prototype.ssn_last_4;
}
/**
 * @record
 */
export function IndividualEntity() { }
if (false) {
    /** @type {?} */
    IndividualEntity.prototype.type;
}
/**
 * @record
 */
export function BusinessEntity() { }
if (false) {
    /** @type {?} */
    BusinessEntity.prototype.type;
    /** @type {?|undefined} */
    BusinessEntity.prototype.additional_owners;
    /** @type {?} */
    BusinessEntity.prototype.business_name;
    /** @type {?|undefined} */
    BusinessEntity.prototype.business_name_kana;
    /** @type {?|undefined} */
    BusinessEntity.prototype.business_name_kanji;
    /** @type {?|undefined} */
    BusinessEntity.prototype.business_tax_id;
    /** @type {?|undefined} */
    BusinessEntity.prototype.business_vat_id;
    /** @type {?|undefined} */
    BusinessEntity.prototype.personal_address;
    /** @type {?|undefined} */
    BusinessEntity.prototype.personal_address_kana;
    /** @type {?|undefined} */
    BusinessEntity.prototype.personal_address_kanji;
    /** @type {?|undefined} */
    BusinessEntity.prototype.tax_id_registrar;
}
/**
 * @record
 */
export function AccountData() { }
if (false) {
    /** @type {?|undefined} */
    AccountData.prototype.legal_entity;
    /** @type {?|undefined} */
    AccountData.prototype.tos_shown_and_accepted;
}
/**
 * @record
 */
export function BankAccountData() { }
if (false) {
    /** @type {?} */
    BankAccountData.prototype.country;
    /** @type {?} */
    BankAccountData.prototype.currency;
    /** @type {?} */
    BankAccountData.prototype.routing_number;
    /** @type {?} */
    BankAccountData.prototype.account_number;
    /** @type {?|undefined} */
    BankAccountData.prototype.account_holder_name;
    /** @type {?|undefined} */
    BankAccountData.prototype.account_holder_type;
}
/**
 * @record
 */
export function PiiData() { }
if (false) {
    /** @type {?} */
    PiiData.prototype.personal_id_number;
}
/**
 * @param {?} account
 * @return {?}
 */
export function isAccount(account) {
    return account === 'account';
}
/**
 * @param {?} accountData
 * @return {?}
 */
export function isAccountData(accountData) {
    return accountData.legal_entity || accountData.tos_shown_and_accepted;
}
/**
 * @param {?} account
 * @return {?}
 */
export function isBankAccount(account) {
    return account === 'bank_account';
}
/**
 * @param {?} bankAccountData
 * @return {?}
 */
export function isBankAccountData(bankAccountData) {
    return ('country' in bankAccountData &&
        'currency' in bankAccountData &&
        'routing_number' in bankAccountData &&
        'account_number' in bankAccountData &&
        (bankAccountData.account_holder_type === 'individual' ||
            bankAccountData.account_holder_type === 'company' ||
            bankAccountData.account_holder_type === undefined));
}
/**
 * @param {?} pii
 * @return {?}
 */
export function isPii(pii) {
    return pii === 'pii';
}
/**
 * @param {?} piiData
 * @return {?}
 */
export function isPiiData(piiData) {
    return 'personal_id_number' in piiData;
}
